doc/nrf/ug_fw_update.rst

```diff
@@ -0,0 +1,311 @@
+
+You can add this feature to your own project and check its functionality as follows:
+
+1. Use this :ref:`Python tool <ug_fw_update_keys_python>` to generate two or more private keys for the application and then extract a public key for each one:
+
+   .. code-block:: console
+
+      python3 nrf/scripts/bootloader/keygen.py --public -i priv.pem -o pub.pem
+
+#. Compile the application and bootloader with the relevant configurations, using only absolute paths:
+
+   .. code-block:: console
+
+      CONFIG_SECURE_BOOT=y
+      CONFIG_SB_SIGNING_KEY_FILE="/path/to/priv_a.pem"
+      CONFIG_SB_PUBLIC_KEY_FILES="/path/to/pub_b.pem,/path/to/pub_c.pem"
+
+   .. note::
+
+      The public key associated with the original private signing key should not be included in the public key list.
+
+#. Program the application to the target development kit and :ref:`check its console output <ug_bootloader_adding_testing>`.
+   With the first firmware version, ``priv_a.pem`` and ``pub_a.pem`` are used for signing and validating the image.
+
+   .. code-block:: console
+
+      *** Booting Zephyr OS build ...  ***
+      Attempting to boot slot 0.
+      Attempting to boot from address 0x9000.
+      Verifying signature against key 0.
+      Hash: 0xda...4f
+      Firmware signature verified.
+      Firmware version 1
+      *** Booting Zephyr OS build ...  ***
+      ...
+
+#. To revoke a single or multiple keys, rebuild the application using the private key associated with a key listed *after* the currently used key in the list.
+   For example, when compiling with the ``priv_c.pem`` key, images signed with ``priv_a.pem`` or ``priv_b.pem`` will no longer boot when uploaded into an image slot.
+
+   Modify the configuration settings as follows, including a new firmware version:
+
+   .. code-block:: console
+
+      CONFIG_BUILD_S1_VARIANT=y
+      CONFIG_SB_SIGNING_KEY_FILE="/path/to/priv_c.pem"
+      CONFIG_FW_INFO_FIRMWARE_VERSION=2
+
+#. Deploy the firmware update.
+   Observe the bootloader checking the hashes of the public keys against the new image, then invalidating the earlier keys:
+
+   .. code-block:: console
+
+      *** Booting Zephyr OS build ...  ***
+      Attempting to boot slot 1.
+      Attempting to boot from address 0x84800.
+      Verifying signature against key 0.
+      Hash: 0xda...4f
+      Public key didn't match, try next.
+      Verifying signature against key 1.
+      Hash: 0x5c...f5
+      Public key didn't match, try next.
+      Verifying signature against key 2.
+      Hash: 0x19...73
+      Invalidating key 0.
+      Invalidating key 1.
+      Firmware signature verified.
+      Firmware version 2
+      Setting monotonic counter (version: 2, slot: 1)
+      *** Booting Zephyr OS build ...  ***
+      ...
+
+Testing
+-------
+
+You can test that the bootloader will no longer boot images signed with the earlier keys by uploading an image signed with one of them.
```

Comment:
This should be a step-based procedure using numbered steps, not steps in separate paragraphs.

Reply:
fixed